examples/acados_python/pendulum_on_cart/solution_sensitivities/policy_gradient_example.py

```diff
@@ -0,0 +1,382 @@
+    ocp.cost.cost_type = "NONLINEAR_LS"
+    ocp.cost.cost_type_e = "NONLINEAR_LS"
+    ocp.cost.W = scipylinalg.block_diag(Q_mat, R_mat)
+    ocp.cost.W_e = Q_mat
+
+    ocp.model.cost_y_expr = vertcat(ocp.model.x[:nx_original], ocp.model.u)
+    ocp.model.cost_y_expr_e = ocp.model.x[:nx_original]
+    ocp.cost.yref = np.zeros((nx_original + nu,))
+    ocp.cost.yref_e = np.zeros((nx_original,))
+
+    # set constraints
+    ocp.constraints.lbu = np.array([-Fmax])
+    ocp.constraints.ubu = np.array([+Fmax])
+    ocp.constraints.idxbu = np.array([0])
+
+    ocp.constraints.x0 = x0[:nx]
+
+    # set options
+    ocp.solver_options.qp_solver = "PARTIAL_CONDENSING_HPIPM"  # FULL_CONDENSING_QPOASES
+    # PARTIAL_CONDENSING_HPIPM, FULL_CONDENSING_QPOASES, FULL_CONDENSING_HPIPM,
+    # PARTIAL_CONDENSING_QPDUNES, PARTIAL_CONDENSING_OSQP, FULL_CONDENSING_DAQP
+    ocp.solver_options.integrator_type = "IRK"  # "DISCRETE"
+    # ocp.solver_options.print_level = 1
+    ocp.solver_options.nlp_solver_type = "SQP"  # SQP_RTI, SQP
+    ocp.solver_options.qp_solver_cond_N = N_horizon
+
+    # set prediction horizon
+    ocp.solver_options.tf = T_horizon
+
+    ocp.solver_options.qp_solver_ric_alg = qp_solver_ric_alg
+    ocp.solver_options.hessian_approx = hessian_approx  # 'GAUSS_NEWTON', 'EXACT'
+    if hessian_approx == 'EXACT':
+        ocp.solver_options.nlp_solver_step_length = 0.0
+        ocp.solver_options.nlp_solver_max_iter = 1
+        ocp.solver_options.qp_solver_iter_max = 200
+        ocp.solver_options.tol = 1e-10
+        # ocp.solver_options.hpipm_mode = 'SPEED_ABS'
+    else:
+        ocp.solver_options.nlp_solver_max_iter = 400
+        ocp.solver_options.tol = 1e-8
+
+    return ocp
+
+
+
+
+def evaluate_hessian_eigenvalues(acados_solver: AcadosOcpSolver):
+    offset = 0
+    min_eigv_total = 1e12
+    min_abs_eigv = 1e12
+    N_horizon = acados_solver.acados_ocp.dims.N
+
+    for i in range(N_horizon+1):
+        hess_block_acados = acados_solver.get_hessian_block(i)
+        nv = hess_block_acados.shape[0]
+        offset += nv
+
+        eigv = np.linalg.eigvals(hess_block_acados)
+        min_eigv = np.min(eigv)
+        min_eigv_total = min(min_eigv, min_eigv_total)
+        min_abs_eigv = min(min_abs_eigv, np.min(np.abs(eigv)))
+
+    # check projected Hessian
+    min_abs_eig_proj_hess = 1e12
+    min_eig_proj_hess = 1e12
+    min_eig_P = 1e12
+    min_abs_eig_P = 1e12
+    for i in range(1, N_horizon):
+        P_mat = acados_solver.get_from_qp_in(i, 'P')
+        B_mat = acados_solver.get_from_qp_in(i, 'B')
+        R_mat = acados_solver.get_from_qp_in(i, 'R')
+        proj_hess_block = R_mat + B_mat.T @ P_mat @ B_mat
+        eigv = np.linalg.eigvals(proj_hess_block)
+        min_eigv = np.min(eigv)
+        min_eig_proj_hess = min(min_eigv, min_eig_proj_hess)
+        min_abs_eig_proj_hess = min(min_abs_eig_proj_hess, np.min(np.abs(eigv)))
```

Comment:
Why do you compute the minimum of the absolute value of the eigenvalues?
As long as they are not numerically very close to zero, the issue is just if there are negative eigenvalues.

Reply:
> As long as they are not numerically very close to zero, the issue is just if there are negative eigenvalues.

I wanted to exactly check if they are close to zero.
I also plotted the sign of the smallest eigenvalue now.